src/app/core/services/authentication.service.ts

```diff
@@ -0,0 +1,126 @@
+import { Injectable } from '@angular/core';
+import { AngularFireAuth } from '@angular/fire/compat/auth';
+import {
+  BehaviorSubject,
+  catchError,
+  from,
+  map,
+  Observable,
+  tap,
+  throwError,
+} from 'rxjs';
+import firebase from 'firebase/compat';
+import FirebaseError = firebase.FirebaseError;
+import { UsersService } from './users.service';
+import { Store } from '@ngrx/store';
+import { removeToken, setToken } from '../states/auth/auth.actions';
+import { resolveError, setErrorMessage } from '../states/error/error.actions';
+import UserCredential = firebase.auth.UserCredential;
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AuthenticationService {
+  loggedIn = new BehaviorSubject<boolean>(false);
+  loggedIn$ = this.loggedIn.asObservable();
+  constructor(
+    private auth: AngularFireAuth,
+    private usersService: UsersService,
+    private store: Store,
+  ) {
+    this.auth.onAuthStateChanged((user) => {
+      if (user) {
+        this.loggedIn.next(true);
+      } else {
+        this.loggedIn.next(false);
+      }
+    });
+  }
+
+  signInWithEmailPassword(
+    email: string,
+    password: string,
+  ): Observable<UserCredential> {
+    return from(this.auth.signInWithEmailAndPassword(email, password)).pipe(
+      tap(
+        (result) =>
+          result.user
+            ?.getIdToken()
+            .then((token) => this.store.dispatch(setToken({ token: token }))),
+      ),
+      tap(() => this.store.dispatch(resolveError({ errorType: 'signIn' }))),
+      catchError((err: FirebaseError) =>
+        throwError(() =>
+          this.store.dispatch(
+            setErrorMessage({
+              error: { error: 'signIn', message: this.convertSignInError(err) },
+            }),
+          ),
+        ),
+      ),
+    );
+  }
+
+  signUpWithEmailPassword(
+    email: string,
+    password: string,
+    displayName: string,
+  ): Observable<void> {
+    return from(this.auth.createUserWithEmailAndPassword(email, password)).pipe(
+      map((result) => {
+        result.user?.updateProfile({ displayName: displayName });
+        this.usersService.create({
+          id: result.user?.uid,
+          collection: { games: [], movies: [], music: [] },
+        });
+        result.user
+          ?.getIdToken()
+          .then((token) => this.store.dispatch(setToken({ token: token })));
```

Comment:
These days you should always prefer usign the syntactic sugar around promise with await/async
This would make your code much more readable:
```
await const token =  result.user?.getIdToken()
this.store.dispatch(setToken({ token: token })));
```

This being said your approach here looks wrong as you combine rxjs, then map a value and inside the map you create a promise chain. Ideally you would use things like switchMap, combine & friends to include other async data. You could get away with a `tap` but the `map` should yield some transformed that which you do not return here.

Reply:
In this case I've now completely removed the token from the store as I wasn't using it anyway (I was using angular/fire's built-in tools to check for authentication). That being said: if I ever have to call a function which returns a promise in an rxjs pipe (like this getIdToken function from angular/fire), how should I deal with it?